File: frontangular/src/app/core/services/caja.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { firstValueFrom } from 'rxjs';
import type {
  CajaDia,
  EstadoCaja,
  GastoCaja,
  IngresoManualCaja,
  ResumenCaja,
  TipoPagoCaja,
} from '../../shared/types';

const BASE = 'http://localhost:3000/api/caja';

@Injectable({ providedIn: 'root' })
export class CajaService {
  private readonly http = inject(HttpClient);

  obtenerEstado(): Promise<EstadoCaja> {
    return firstValueFrom(this.http.get<EstadoCaja>(`${BASE}/estado`));
  }

  abrir(montoInicial: number, observaciones?: string): Promise<CajaDia> {
    return firstValueFrom(
      this.http.post<CajaDia>(`${BASE}/abrir`, { montoInicial, observaciones }),
    );
  }

  obtenerResumen(cajaDiaId: string): Promise<ResumenCaja> {
    return firstValueFrom(this.http.get<ResumenCaja>(`${BASE}/resumen/${cajaDiaId}`));
  }

  cerrar(cajaDiaId: string): Promise<CajaDia> {
    return firstValueFrom(this.http.post<CajaDia>(`${BASE}/cerrar/${cajaDiaId}`, {}));
  }

  registrarGasto(concepto: string, monto: number, tipoPago: TipoPagoCaja): Promise<GastoCaja> {
    return firstValueFrom(
      this.http.post<GastoCaja>(`${BASE}/gastos`, { concepto, monto, tipoPago }),
    );
  }

  eliminarGasto(id: string): Promise<void> {
    return firstValueFrom(this.http.delete<void>(`${BASE}/gastos/${id}`));
  }

  registrarIngresoManual(
    concepto: string,
    monto: number,
    tipoPago: TipoPagoCaja,
  ): Promise<IngresoManualCaja> {
    return firstValueFrom(
      this.http.post<IngresoManualCaja>(`${BASE}/ingresos-manuales`, {
        concepto,
        monto,
        tipoPago,
      }),
    );
  }

  eliminarIngresoManual(id: string): Promise<void> {
    return firstValueFrom(this.http.delete<void>(`${BASE}/ingresos-manuales/${id}`));
  }

  historial(): Promise<CajaDia[]> {
    return firstValueFrom(this.http.get<CajaDia[]>(`${BASE}/historial`));
  }
}

File: applavadero/lib/core/services/caja_service.dart
import '../api/api_client.dart';
import '../api/api_endpoints.dart';
import '../models/caja.dart';

class CajaService {
  final ApiClient _api;
  CajaService(this._api);

  Future<EstadoCaja> getEstado() async {
    final res = await _api.get(ApiEndpoints.cajaEstado);
    return EstadoCaja.fromJson(res as Map<String, dynamic>);
  }

  Future<CajaDia> abrir(double montoInicial, {String? observaciones}) async {
    final data = <String, dynamic>{'montoInicial': montoInicial};
    if (observaciones != null && observaciones.isNotEmpty) {
      data['observaciones'] = observaciones;
    }
    final res = await _api.post(ApiEndpoints.cajaAbrir, data: data);
    return CajaDia.fromJson(res as Map<String, dynamic>);
  }

  Future<GastoCaja> registrarGasto(
      String concepto, double monto, String tipoPago) async {
    final res = await _api.post(ApiEndpoints.cajaGasto,
        data: {'concepto': concepto, 'monto': monto, 'tipoPago': tipoPago});
    return GastoCaja.fromJson(res as Map<String, dynamic>);
  }

  Future<void> eliminarGasto(String id) async {
    await _api.delete(ApiEndpoints.cajaGastoEliminar(id));
  }

  Future<IngresoManualCaja> registrarIngreso(
      String concepto, double monto, String tipoPago) async {
    final res = await _api.post(ApiEndpoints.cajaIngresoManual, data: {
      'concepto': concepto,
      'monto': monto,
      'tipoPago': tipoPago,
    });
    return IngresoManualCaja.fromJson(res as Map<String, dynamic>);
  }

  Future<void> eliminarIngreso(String id) async {
    await _api.delete(ApiEndpoints.cajaIngresoEliminar(id));
  }

  Future<CajaDia> cerrar(String id) async {
    final res = await _api.post(ApiEndpoints.cajaCerrar(id));
    return CajaDia.fromJson(res as Map<String, dynamic>);
  }

  Future<CajaDia> reabrir(String id) async {
    final res = await _api.post(ApiEndpoints.cajaReabrir(id));
    return CajaDia.fromJson(res as Map<String, dynamic>);
  }

  Future<ResumenCaja> getResumen(String id) async {
    final res = await _api.get(ApiEndpoints.cajaResumen(id));
    return ResumenCaja.fromJson(res as Map<String, dynamic>);
  }
}
